### PR TITLE
Fix polyline stroke

### PR DESCRIPTION
**Pull request by floppyhammer**
Fix incorrect polyline stroking.

Before
image

After
image

Other shapes
image

### CODE REVIEW

**Comment by Calinou** [comment]
Does this fully resolve https://github.com/godotengine/godot/issues/56320, or only partially?

**Comment by floppyhammer** [comment]
@Calinou This partially fixes https://github.com/godotengine/godot/issues/56320. But in my opinion, how "Snap 2D Transforms To Pixel" and "Snap 2D Vertices To Pixel" options behave can hardly be sorted as issues. As for https://github.com/godotengine/godot/pull/62236, I suppose this PR solves the ununiform line width issue mentioned there.

In addition, I'm trying to implement this for anti-aliased polylines as well but having problem testing it. draw_polyline() doesn't accept `Color` now?

image

**Comment by Calinou** [comment]
image

Maybe color constants are broken in some cases. Try using a fixed color like `Color(1, 0, 0)` instead of `Color.RED`.

File: servers/rendering/renderer_canvas_cull.cpp [hunk 1]
```diff
@@ -1014,30 +1014,44 @@ void RendererCanvasCull::canvas_item_add_polyline(RID p_item, const Vector<Point
...
+					// No previous line for the first line.
 					prev_t = t;
 				}
 			}
 
-			j2 = j + 1;
+			// Clamp the argument that we're going to pass to acos() within [-1, 1].
+			real_t dot_product = t.dot(prev_t);
+			dot_product = dot_product < -1.0 ? -1.0 : dot_product;
+			dot_product = dot_product > 1.0 ? 1.0 : dot_product;
```
Comment (clayjohn) [inline servers/rendering/renderer_canvas_cull.cpp:1033]:
```suggestion
			dot_product = CLAMP(dot_product, -1.0, 1.0);
```

**Comment by floppyhammer** [comment]
@Calinou It doesn't work. It seems the parameter color is now a `RID` instead of a `Color`.

image

**Comment by floppyhammer** [comment]
Updated for the anti-aliased case.

Before
Screenshot 2022-07-27 125848


After
Screenshot 2022-07-27 130723

**Comment by Calinou** [comment]
> It doesn't work. It seems the parameter color is now a RID instead of a Color.

It's likely because the Vector4 addition broke things.

**Comment by MewPurPur** [comment]
Is this still useful after #62236 was merged?

**Comment by floppyhammer** [comment]
@MewPurPur I believe so.

After https://github.com/godotengine/godot/pull/62236 was merged:
image

This PR:
image

However, the merge conflicts are quite an issue now. It might not be worth it.

**Comment by dalexeev** [comment]
I'm assuming #62236 and #71679 completely fixed polyline rendering (I could be wrong, let me know if there are any other useful changes in this PR).

Also note that the rendering of 2D collision shapes and `CollisionPolygon2D` has been slightly changed in #72302. It's not really correct to use collision shapes to debug common methods like `canvas_item_add_polyline`, you should create a test project, which will only use the debugged method (examples: [comment 1](https://github.com/godotengine/godot/pull/41239#issuecomment-1335397317), [comment 2](https://github.com/godotengine/godot/pull/69851#issuecomment-1346173276)).

**Pull request closed without merging**